Source/Router.swift

```diff
@@ -210,14 +213,31 @@ public final class Router {
 
         let delayTime = dispatch_time(DISPATCH_TIME_NOW, Int64(latency * Double(NSEC_PER_SEC)))
         dispatch_after(delayTime, dispatch_get_main_queue()) {
-            didFinishLoading(server)
+            [weak self] in
+            // before reporting "finished", check if request has been canceled in the meantime
+            guard let strongSelf = self else {
+                return
+            }
+            if let serverIndex = strongSelf.canceledRequests.indexOf(server) {
+                // remove server from the list of "canceled requests" and DO NOT send notification(s)
+                strongSelf.canceledRequests.removeAtIndex(serverIndex)
+            }
+            else {
+                didFinishLoading(server)
+            }
+        }
+    }
+
+    func stopLoading(server: NSURLProtocol) {
+        if canceledRequests.contains(server) == false {
```

Comment:
Not familiar with NSURLProtocol (Joan implemented it) , looks like an instance of KakapoServer is created for every request? If yes then we could just have a flag in the Server (cancelled) and check it in startLoading